src/parser/jobs/sam/modules/Shoha.tsx

```diff
@@ -0,0 +1,272 @@
+import Combatants from 'parser/core/modules/Combatants'
+import {ComboEvent} from 'parser/core/modules/Combos'
+import Suggestions, {SEVERITY, TieredSuggestion} from 'parser/core/modules/Suggestions'
+
+const GENERATORS = {
+	[ACTIONS.HIGANBANA.id]: 1,
+	[ACTIONS.KAESHI_HIGANBANA.id]: 1,
+	[ACTIONS.TENKA_GOKEN.id]: 1,
+	[ACTIONS.KAESHI_GOKEN.id]: 1,
+	[ACTIONS.MIDARE_SETSUGEKKA.id]: 1,
+	[ACTIONS.KAESHI_SETSUGEKKA.id]: 1,
+}
+
+const SPENDERS = {
+	[ACTIONS.SHOHA.id]: 3,
+}
+
+const MAX_STACKS = 3
+
+class StackState {
+	t?: number
+	y?: number
+}
+
+export default class Shoha extends Module {
+	static handle = 'shoha'
+	static title = t('sam.shoha.title')`Meditation Timeline`
+	static displayMode = DISPLAY_MODE.FULL
+
+	private stacks = 0
+	private shohaUses = 0
+	private stackHistory: StackState[] = []
+	private wasteBySource = {
+		[ACTIONS.HIGANBANA.id]: 0,
+		[ACTIONS.KAESHI_HIGANBANA.id]: 0,
+		[ACTIONS.TENKA_GOKEN.id]: 0,
+		[ACTIONS.KAESHI_GOKEN.id]: 0,
+		[ACTIONS.MIDARE_SETSUGEKKA.id]: 0,
+		[ACTIONS.KAESHI_SETSUGEKKA.id]: 0,
+		[ACTIONS.RAISE.id]: 0,
+	}
+	private leftoverStacks = 0
+	private totalGeneratedStacks = 0 // Keep track of the total amount of generated stacks over the fight
+
+	@dependency private checklist!: Checklist
+	@dependency private suggestions!: Suggestions
+	@dependency private combatants!: Combatants
+
+	protected init() {
+		this.addHook('init', this.pushToHistory)
+		this.addHook(
+			'cast',
+			{
+				by: 'player',
+				abilityId: Object.keys(GENERATORS).map(Number),
+			},
+			this.onGenerator,
+		)
+		this.addHook(
+			'cast',
+			{
+				by: 'player',
+				abilityId: Object.keys(SPENDERS).map(Number),
+			},
+			this.onSpender,
+		)
+		this.addEventHook([
+			'applybuff', 'applybuffstack'],
+			{
+			to: 'player',
+			abilityId: STATUSES.MEDITATION.id,
+			}, this.onMeditateGain)
+
+		this.addHook('death', {to: 'player'}, this.onDeath)
+		this.addHook('complete', this.onComplete)
+	}
+
+	private onGenerator(event: CastEvent) {
+		if (this.stacks === MAX_STACKS) {
+		const abilityId = event.ability.guid
+		const generatedStacks = GENERATORS[abilityId]
+
+		this.addGeneratedStackAndPush(generatedStacks, abilityId)
+	}
+	}
+
+		// This is exclusively for stacks gained via Meditate channel. and not any stacks gained from GCD casts
+	private onMeditateGain() {
+			this.stacks++
+			this.totalGeneratedStacks++
+
+			// just a safety net
+			if (this.stacks > MAX_STACKS) {
+				this.stacks = MAX_STACKS
+			}
+
+			this.pushToHistory()
+	}
+
+	private addGeneratedStackAndPush(generatedStacks: number, abilityId: number) {
+		this.stacks += generatedStacks
+		this.totalGeneratedStacks += generatedStacks
+		if (this.stacks > MAX_STACKS) {
+			const waste = this.stacks - MAX_STACKS
+			this.wasteBySource[abilityId] += waste
+			this.stacks = MAX_STACKS
+		}
+
+		this.pushToHistory()
+	}
+
+	private onSpender(event: CastEvent) {
+		this.stacks = this.stacks - SPENDERS[event.ability.guid]
```

Comment:
If max is 3 and Shoha is 3, why not just reset it, or trying to catch broken logs (which you don't do in this PR)?

Reply:
I can change that. A lot of this code has some useless future proofing in case they add in ways to spend stacks without being at cap so I could just throw the new spenders/generators into the const groups for each. in hindsight it just makes the code look more chaotic.